### DISCUSSION TITLE
Exporting and Loading nested Pydantic models

### DISCUSSION THREAD

**Discussion by twhughes** [pydantic/pydantic#3232] [Question]
I have a simple pydantic model with nested data structures (and inheritance). I want to be able to simply save and load instances of this model as .json file.

All models inherit from a Base class with simple configuration.
```python
class Base(pydantic.BaseModel):
    class Config:
        extra = 'forbid'   # forbid use of extra kwargs
```
There are some simple data models with inheritance
```python
class Thing(Base):
    thing_id: int

class SubThing(Thing):
    name: str
```
And a `Container` class, which holds a `Thing`
```python
class Container(Base):
    thing: Thing
```
I can create a Container instance and save it as .json
```python
# make instance of container
c = Container(
    thing = SubThing(
        thing_id=1,
        name='my_thing')
)

json_string = c.json(indent=2)
print(json_string)

"""
{
  "thing": {
    "thing_id": 1,
    "name": "my_thing"
  }
}
"""
```
but the json string does not specify that the thing field was constructed using a SubThing. As such, when I try to load this string into a new Container instance, I get an error.
```python
c = Container.parse_raw(json_string)
print(c)
"""
Traceback (most recent call last):
  File "...", line 36, in <module>
    c = Container.parse_raw(json_string)
  File "pydantic/main.py", line 601, in pydantic.main.BaseModel.parse_raw
  File "pydantic/main.py", line 578, in pydantic.main.BaseModel.parse_obj
  File "pydantic/main.py", line 406, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 1 validation error for Container
thing -> name
  extra fields not permitted (type=value_error.extra)
"""
```
Is there a simple way to save the Container instance while retaining information about the thing class type such that I can reconstruct the initial Container instance reliably? I would like to avoid pickling the object if possible.

**Comment by MikesRND** [comment]
This appears closely related to the same question that I have: #3293.

**Reply by twhughes** [reply]
Yea seems like it, for now I ended up just adding 

```python
type: Literal['my_type'] = 'my_type'
```

To each of the subclasses, which inserts the `type` field into each of the subclasses and tells pedantic how to load them.

It would be nice if there was a global config in pydantic to append this `type` to each class automatically.

**Reply by MikesRND** [reply]
Thanks, @twhughes.  That doesn't seem to work for me.  Would I then need to override parse__obj method to look at this field?

**Reply by twhughes** [reply]
Hm. Did you change "my_type" so it's different for each subclass?

**Reply by twhughes** [reply]
I just commented directly on your question, but I think you'd also need to change the type annotation of the value of `stuff` to `Union[SubclassA, SubclassB]` to be explicit

**Reply by twhughes** [reply]
Or `Union[SubclassA, SubclassB, MainClass]` if you want to include `MainClass` (I think it will look for matches in `Union` from left to right)

**Reply by MikesRND** [reply]
Thanks again @twhughes.  I did get it to work using your suggestion, although it seems a little like a back door, and maybe not a coding pattern I could trust in production code.

This feels like it should be a an issue or feature reuqest.

**Comment by MikesRND** [comment]
I also found a suggestion by @PrettyWood in response to this issue:

* https://github.com/samuelcolvin/pydantic/issues/2177

It explicitly registers the subclasses, which is something that is probably required for robustness.  Hopefully this kind of thing can get built in.

**Reply by twhughes** [reply]
Thanks, this is similar to what I tried earlier actually, although I never figured out how to generalize it such that it worked automatically for all classes that inherit from my BaseModel where this is defined.

See similar discussion on stack overflow that I posted also.
https://stackoverflow.com/questions/69322097/distinguishing-between-pydantic-models-with-same-fields
